CORE-8691: Delay Closing SandboxGroupContext

jujoramos opened the pull request:
Prevent automatically closing SandboxGroupContexts as soon as they are
evicted from the cache since they might still be in use. Instead, keep
a WeakReference to an internal wrapper and only close the evicted
SandboxGroupContext once all strong references to the wrapper are gone.

- Fix adapted from https://github.com/corda/corda/commit/4316838.

jujoramos committed:
CORE-8691: Delay Closing SandboxGroupContext

Prevent automatically closing SandboxGroupContexts as soon as they are
evicted from the cache since they might still be in use. Instead, keep
a WeakReference to an internal wrapper and only close the evicted
SandboxGroupContext once all strong references to the wrapper are gone.

- Fix adapted from https://github.com/corda/corda/commit/4316838.

jujoramos committed:
Implement AutoClosable

JamesHR3 commented on components/virtual-node/sandbox-group-context-service/src/main/kotlin/net/corda/sandboxgroupcontext/service/impl/SandboxGroupContextCacheImpl.kt:
```diff
@@ -5,34 +5,106 @@ import com.github.benmanes.caffeine.cache.Caffeine
+    )
 
     override fun remove(virtualNodeContext: VirtualNodeContext) {
+        purgeExpiryQueue()
+
         contexts.invalidate(virtualNodeContext)
```
Should we do these the other way around in case we can immediately remove it, or does that lead to problems?

jujoramos replied:
Invalidated entries are closed straight away (they are not added to the internal `toBeClosed` structure) because they are not "evictions", so it shouldn't make any difference: 
```
.removalListener { key, context, cause ->
                if (cause.wasEvicted()) {
                  ...
                } else {
                    logger.info(
                        "Closing ${key!!.sandboxGroupType} sandbox for " +
                                "${key.holdingIdentity.x500Name} [${cause.name}]"
                    )

                    (context?.wrappedSandboxGroupContext as? AutoCloseable)?.close()
                }
```

JamesHR3 replied:
Great!

JamesHR3 commented on components/virtual-node/sandbox-group-context-service/src/test/kotlin/net/corda/sandboxgroupcontext/impl/SandboxGroupContextCacheTest.kt:
```diff
@@ -5,134 +5,193 @@ import net.corda.sandboxgroupcontext.VirtualNodeContext
 
-        assertThat(retrievedContext).isSameAs(sandboxContext1)
+    @Test
+    fun `sandboxes of different types do not trigger close on eviction of other sandbox types`() {
```
Is this test valid? We're not partitioning the cache by sandbox type right?

jujoramos replied:
We're not partitioning the cache, no, but I still considered it as a valid test to have checked in (specially for the combined worker situation), that's the only reason why I copied it over from your previous `PR`. Happy to remove it if you consider it necessary, though.

JamesHR3 replied:
The change I made before was to partition the cache by sandbox type, so adding a new flow sandbox should never trigger persistence sandbox eviction for example. That's why I added that test. I don't think that's true for your code though? So I think this could (should?) fail.

jujoramos replied:
The cache has size `2` and we're adding `3` different `SandboxGroupContext`s to it, two of them will be evicted and the other will remain in the cache, but no context will be closed (that's what the test asserts, so it shouldn't fail). I'll remove the test anyway 👍 .

jujoramos committed:
Remove unnecessary test

chrisr3 commented on components/virtual-node/sandbox-group-context-service/src/main/kotlin/net/corda/sandboxgroupcontext/service/impl/SandboxGroupContextCacheImpl.kt:
```diff
@@ -41,4 +113,22 @@ internal class SandboxGroupContextCacheImpl(override val capacity: Long): Sandbo
+        override fun <T : Any> get(key: String, valueType: Class<out T>) =
+            wrappedSandboxGroupContext.get(key, valueType)
+
+        override fun close() = wrappedSandboxGroupContext.close()
```
Isn't this dangerous? In fact, why make `SandboxGroupContextWrapper` implement `AutoCloseable` at all when we cannot allow anyone to invoke its `close()` method? Only the _underlying_ context should be closed, and only when its wrapper has already been garbage-collected.

jujoramos replied:
It wasn't `AutoCloseable` originally, but [there's code](https://github.com/corda/corda-runtime-os/blob/release/os/5.0/components/virtual-node/sandbox-group-context-service/src/integrationTest/kotlin/net/corda/sandboxgroupcontext/test/VirtualNodeService.kt#L85-L97) within the `VirtualNodeService` that relies on manually unloading and closing the `SandboxGroupContext`.
Apparently you added this code back in February?, do you think it needs to be removed or should the wrapper somehow remove itself from the cache when somebody manually closes it (as it's not valid anymore for anyone else)?.

jujoramos replied:
The code that manually closes the context is only used within tests (`CustomCryptoDigestTests`), I'll need to update things so the sandboxes are removed from the cache instead of manually closed by the `VirtualNodeService`.

chrisr3 commented on components/virtual-node/sandbox-group-context-service/src/main/kotlin/net/corda/sandboxgroupcontext/service/impl/SandboxGroupContextCacheImpl.kt:
```diff
@@ -5,34 +5,106 @@ import com.github.benmanes.caffeine.cache.Caffeine
-            logger.info("Caching {} sandbox for: {} (cache size: {})",
-                virtualNodeContext.sandboxGroupType, virtualNodeContext.holdingIdentity.x500Name, contexts.estimatedSize())
-            createFunction(virtualNodeContext)
+            logger.info(
+                "Caching ${virtualNodeContext.sandboxGroupType} sandbox for " +
+                        "${virtualNodeContext.holdingIdentity.x500Name} (cache size: ${contexts.estimatedSize()})"
+            )
```
Please revert these log messages back to using SLF4J placeholders, which exist precisely to avoid lots of unnecessary string concatenation. Just because Kotlin has made all that concatenation "more convenient" to write still doesn't make it a Good Idea.

jujoramos replied:
👍

chrisr3 commented on components/virtual-node/sandbox-group-context-service/src/main/kotlin/net/corda/sandboxgroupcontext/service/impl/SandboxGroupContextCacheImpl.kt:
```diff
@@ -5,34 +5,106 @@ import com.github.benmanes.caffeine.cache.Caffeine
+
+                    purgeExpiryQueue()
+                } else {
+                    logger.info(
+                        "Closing ${key!!.sandboxGroupType} sandbox for " +
+                                "${key.holdingIdentity.x500Name} [${cause.name}]"
+                    )
```
You didn't need to change these log messages, they were better using SLF4J placeholders.

jujoramos replied:
👍

chrisr3 commented on components/virtual-node/sandbox-group-context-service/src/test/kotlin/net/corda/sandboxgroupcontext/impl/SandboxGroupContextCacheTest.kt:
```diff
@@ -5,134 +5,157 @@ import net.corda.sandboxgroupcontext.VirtualNodeContext
+    private lateinit var idAlice: HoldingIdentity
+    private lateinit var vNodeContext1: VirtualNodeContext
+
+    @BeforeEach
+    fun setUp() {
+        idBob = createTestHoldingIdentity("CN=Bob, O=Bob Corp, L=LDN, C=GB", "group")
+        idAlice = createTestHoldingIdentity("CN=Alice, O=Alice Corp, L=LDN, C=GB", "group")
+
+        vNodeContext1 = VirtualNodeContext(
+            holdingIdentity = idBob,
+            serviceFilter = "filter",
+            sandboxGroupType = SandboxGroupType.FLOW,
+            cpkFileChecksums = setOf(SecureHash.parse("SHA-256:1234567890"))
+        )
+    }
 
```
Do these mocks test the behaviour during _actual_ garbage collection? Because that's the behaviour that matters.

jujoramos replied:
I haven't found a reliable way of making the JVM invoke the garbage collector so, instead of that, I'm "simulating" the behaviour by invoking `WeakReference.enqueue()` from within the relevant tests.

chrisr3 replied:
As a rule, you null out all string references, invoke `System.gc()` and then "wait a bit"... :wink:

jujoramos replied:
The `System.gc()` call just suggests the JVM to trigger a garbage collection execution, there's no guarantee that it's going to actually do it when we want it to. Aside from that, it might take more than just one GC cycle for the JVM to effectively update the `ReferenceQueue`. I'd rather keep the current approach and use `enqueue()` directly, it'll reduce the chance of flakiness.

chrisr3 replied:
> The System.gc() call just suggests the JVM to trigger a garbage collection execution, there's no guarantee that it's going to actually do it when we want it to.

Yes, that's why I said you need to "wait a bit" :no_mouth:.

The biggest risk with using `WeakReference` is that you might accidentally miss a strong reference which prevents the object from ever  being enqueued in practice. I don't see how you can test that using mocks.

jujoramos replied:
I've updated the test to continuously force a GC execution followed by a cache eviction, the `ReferenceQueue` will be eventually updated and there should be no flakiness.

chrisr3 reviewed (changes requested):
I don't think `SandboxGroupContextWrapper` should implement `AutoCloseable` precisely because we don't want anyone closing it and destroying the underlying sandbox.

I was also using SLF4J placeholders in those log messages _for a reason_.

jujoramos committed:
Revert back to SLF4J placeholders

jujoramos committed:
Review Feedback:

- Remove AutoCloseable to prevent manually unloading cached sandboxes.
- Update test to verify interaction with the GC instead of using mocks.

jujoramos committed:
Improve test

jujoramos committed:
Add missing @Suppress(ExplicitGarbageCollectionCall)

chrisr3 commented on components/virtual-node/sandbox-group-context-service/src/test/kotlin/net/corda/sandboxgroupcontext/impl/SandboxGroupContextCacheTest.kt:
```diff
@@ -1,138 +1,184 @@
\ No newline at end of file
+}
+
+@Suppress("ExplicitGarbageCollectionCall")
+fun forceGarbageCollectionExecution() {
+    var obj: String? = String()
+    val ref = WeakReference(obj)
+    obj = null
+    assertThat(obj as? Any).isNull()  // Annoying check to prevent compilation failure due to unused var
+    System.gc()
+
+    assertThat(ref.get())
+        .withFailMessage("garbage collector did not run")
+        .isNull()
+}
```
This function is neither safe nor necessary, and appears to assume that garbage-collection is an instant "all or nothing" kind of operation anyway!!? _In my experience,_ it is sufficient to invoke `System.gc()` and then use something like:
```kotlin
eventually(duration = of(60, SECONDS)) {
    // Has sandbox been garbage-collected yet?
}
```

jujoramos replied:
You're right about the the assertion, I forgot to wrap all invocations to `forceGarbageCollectionExecution()` within a `eventually` clause, thanks for catching that.
I've dealt with `System.gc()` in the past and, according to what I recall from those situations, _it is not sufficient to just call it once and wait for the gc to run_. Considering that within this test we're wrapping the call within an outer `eventually` call, though, things should be fine. I've removed the `forceGarbageCollectionExecution()` function entirely as requested.

jujoramos committed:
"Review feedback"

chrisr3 commented on components/virtual-node/sandbox-group-context-service/src/main/kotlin/net/corda/sandboxgroupcontext/service/impl/SandboxGroupContextCacheImpl.kt:
```diff
@@ -41,4 +123,20 @@ internal class SandboxGroupContextCacheImpl(override val capacity: Long): Sandbo
+     * Wrapper around [CloseableSandboxGroupContext], solely used to keep a [WeakReference] to every instance and only
+     * invoke [CloseableSandboxGroupContext.close] on cache eviction when all strong references are gone.
+     */
+    internal class SandboxGroupContextWrapper(
+        val wrappedSandboxGroupContext: CloseableSandboxGroupContext
+    ) : SandboxGroupContext {
+
+        override fun <T : Any> get(key: String, valueType: Class<out T>) =
+            wrappedSandboxGroupContext.get(key, valueType)
+
+        override val sandboxGroup = wrappedSandboxGroupContext.sandboxGroup
+
+        override val virtualNodeContext: VirtualNodeContext = wrappedSandboxGroupContext.virtualNodeContext
+    }
```
FYI, Kotlin allows you to rewrite this class as:
```kotlin
internal class SandboxGroupContextWrapper(
    private val wrappedSandboxGroupContext: SandboxGroupContext
) : SandboxGroupContext by wrappedSandboxGroupContext
```
and we probably _do_ want `wrappedSandboxGroupContext` to be `private`.

jujoramos replied:
Thanks for the tip, new to `kotlin` here :-). That said, we need access to the `wrappedSandboxGroupContext` from within the outer class in order to close it, so I don't think we can make it `private` (I've changed it to `internal`, though).

chrisr3 replied:
> That said, we need access to the `wrappedSandboxGroupContext` from within the outer class in order to close it, so I don't think we can make it `private` (I've changed it to `internal`, though).

Fair enough, it was the potential for `public` access that was worrying me, even if `internal` scope is only a figment of Kotlin's imagination... Come to think of it, could we have
```kotlin
private class SandboxGroupContextWrapper
```
because I think a `private` class in Kotlin is equivalent to "package private" in Java.

jujoramos replied:
That might work... let me check how can I accommodate the tests with this change.

chrisr3 replied:
I'm hoping the tests only see `SandboxGroupContext`, and don't care which class is actually implementing it.

jujoramos replied:
They will now 💯

jujoramos committed:
Review Feedback

chrisr3 commented on components/virtual-node/sandbox-group-context-service/src/test/kotlin/net/corda/sandboxgroupcontext/impl/SandboxGroupContextCacheTest.kt:
```diff
@@ -1,138 +1,172 @@
+
+        // Remove the strong reference to the context, so it can be garbage collected
+        contextStrongRef = null
+        assertThat(contextStrongRef as? Any).isNull() // Annoying check to prevent compilation failure due to unused var
```
You can always use `@Suppress("unused")` or similar to remove the compiler warning.

chrisr3 commented on components/virtual-node/sandbox-group-context-service/src/test/kotlin/net/corda/sandboxgroupcontext/impl/SandboxGroupContextCacheTest.kt:
```diff
@@ -1,138 +1,172 @@
+
+        // Remove the strong reference to the context, so it can be garbage collected
+        contextStrongRef = null
+        assertThat(contextStrongRef as? Any).isNull() // Annoying check to prevent compilation failure due to unused var
```
Try
```kotlin
// Remove the strong reference to the context, so it can be garbage collected
@Suppress("unused_value")
contextStrongRef = null
```

jujoramos replied:
👍

jujoramos committed:
Review feedback: proper encapsulation and testing

chrisr3 reviewed (approved):
LGTM
